java/ql/src/experimental/Security/CWE/CWE-176/RegexAndStrManipulation.qll

```diff
@@ -0,0 +1,59 @@
+/** Provides methods related to regular expression matching and String manipulation. */
+
+import java
+
+/**
+ * The class `java.util.regex.Pattern`.
+ */
+class Pattern extends RefType {
+  Pattern() { this.hasQualifiedName("java.util.regex", "Pattern") }
+}
+
+/**
+ * The method `compile` of `java.util.regex.Pattern`, or one of its subtypes.
+ */
+class PatternCompileMethod extends Method {
+  PatternCompileMethod() {
+    this.getDeclaringType().getASupertype*() instanceof Pattern and
+    this.hasName("compile")
+  }
+}
+
+/**
+ * The method `matches` of `java.util.regex.Pattern`, or one of its subtypes.
+ */
+class PatternMatchMethod extends Method {
+  PatternMatchMethod() {
+    this.getDeclaringType().getASupertype*() instanceof Pattern and
+    this.hasName("matches")
+  }
+}
+
+/**
+ * The method `matcher` of `java.util.regex.Pattern`, or one of its subtypes.
+ */
+class PatternMatcherMethod extends Method {
+  PatternMatcherMethod() {
+    this.getDeclaringType().getASupertype*() instanceof Pattern and
+    this.hasName("matcher")
+  }
+}
+
+/**
+ * The methods like `matches` of `java.lang.String`, or one of its subtypes.
+ */
+class StringMethod extends Method {
+  StringMethod() {
+    this.getDeclaringType().getASupertype*() instanceof TypeString and
```

Comment:
```suggestion
 * A `java.lang.String` method other than `String.valueOf`.
 */
class StringMethod extends Method {
  StringMethod() {
    this.getDeclaringType() instanceof TypeString and
```

`String` is final. Do you really mean to select every single method except `valueOf`? What is the rationale for picking these? If you really meant any method declared on `String` just drop the `hasName` constraint entirely.

Reply:
I expect any string manipulation method call to be bypassable. Indeed, it should be dropped.

Done via https://github.com/github/codeql/pull/12995/commits/46137523c71f60d2359845302c13b484246a6933

Reply:
Let me dig more please...

Reply:
Yes, indeed, the `String` class is `final`. Can't have a child for sure.

Done via https://github.com/github/codeql/pull/12995/commits/d80899a03fc814a7aa0d487a03ddef18bb8666b0

Reply:
Regarding the tests , I have added in https://github.com/github/codeql/pull/12995/commits/07a16ccc87eda7c0811ea8f5cf5f17d55f85cc1f a good one, and a bad one. The Bad case requires a bit of update for the query ...

Reply:
I'm working on it ...

Reply:
In https://github.com/github/codeql/pull/12995/commits/7ecc9166db30828b2ec0c6d03508dc05086ab2b8, I have added a barrier guard for the Regex matcher use.

Reply:
Finally, updated the `.expected` file in https://github.com/github/codeql/pull/12995/commits/a9a2f7a1c14f23005a4c58946805535eec1ff981.